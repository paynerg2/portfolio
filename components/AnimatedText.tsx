import { motion, AnimatePresence, useAnimation } from 'framer-motion';
import { useEffect } from 'react';
import { useInView } from 'react-intersection-observer';

interface Props {
    children: string;
}
export const AnimatedText = ({ children }: Props) => {
    const controls = useAnimation();
    const [ref, inView] = useInView();

    // Start the animation when scrolled into view
    useEffect(() => {
        if (inView) {
            controls.start('after');
        }
    }, [controls, inView]);

    return (
        <AnimatePresence>
            <motion.h2
                ref={ref}
                style={{ height: '30px' }}
                variants={letterContainerVariants}
                initial={'before'}
                animate={controls}
                exit={'before'}
                key={children}
                aria-label={children}
            >
                {children.split('').map((word: string, index: number) => (
                    <div key={`word-${word}-${index}`} style={{ display: 'inline-block' }}>
                        {Array.from(word).map((letter, index) => (
                            <motion.span
                                key={`${index}-${letter}`}
                                style={{
                                    position: 'relative',
                                    display: 'inline-block',
                                    width: 'auto',
                                }}
                                variants={letterVariants}
                            >
                                {letter === ' ' ? '\u00A0' : letter}
                            </motion.span>
                        ))}
                        {'\u00A0'}
                    </div>
                ))}
            </motion.h2>
        </AnimatePresence>
    );
};

const letterContainerVariants = {
    before: {
        transition: {
            staggerChildren: 0.015,
        },
    },
    after: {
        transition: {
            staggerChildren: 0.03,
        },
    },
};

const letterVariants = {
    before: {
        opacity: 0,
        y: 20,
        transition: {
            type: 'spring',
            damping: 12,
            siffness: 200,
        },
    },
    after: {
        opacity: 1,
        y: 0,
        transition: {
            type: 'spring',
            damping: 12,
            stiffness: 200,
        },
    },
};
