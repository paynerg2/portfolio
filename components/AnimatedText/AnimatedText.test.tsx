import { render, screen } from '../../test/test-utils';
import { AnimatedText } from './AnimatedText';

const mock = function () {
    return {
        observe: jest.fn(),
        unobserve: jest.fn(),
        disconnect: jest.fn(),
    };
};
window.IntersectionObserver = jest.fn().mockImplementation(mock);

describe('Animated Text Component', () => {
    it('renders the component', () => {
        const component = render(<AnimatedText text="test" animationType={0} />);
        expect(component).toMatchSnapshot();
    });

    it('renders the given text', () => {
        const text = 'admndj7843mdfk8as;??*';
        render(<AnimatedText text={text} animationType={0} />);

        const renderedText = screen.getByLabelText(text);
        expect(renderedText).toBeInTheDocument();
    });
});
