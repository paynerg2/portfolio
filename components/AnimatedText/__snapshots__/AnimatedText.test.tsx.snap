// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animated Text Component renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-dark"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        aria-label="test"
        class="chakra-container css-1ulwkta"
      >
        <span
          style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
        >
          t
        </span>
        <span
          style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
        >
          e
        </span>
        <span
          style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
        >
          s
        </span>
        <span
          style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
        >
          t
        </span>
         
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      aria-label="test"
      class="chakra-container css-1ulwkta"
    >
      <span
        style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        t
      </span>
      <span
        style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        e
      </span>
      <span
        style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        s
      </span>
      <span
        style="position: relative; display: inline-block; width: auto; opacity: 0; transform: translateY(20px) translateZ(0);"
      >
        t
      </span>
       
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
